### PR TITLE
v0.12 regression: Fix registration of children for Block

### PR DESCRIPTION
**Pull request by leezu**
If the attribute is already declared (e.g. as None), the registration would fail.
Consider the following example:

```
class BaseClass(gluon.HybridBlock):
    some_block = None

class Implementation(BaseClass):
    def __init__(self):
        some_block = SomeBlock()

```

The regression was introduced by https://github.com/apache/incubator-mxnet/issues/7667

### CODE REVIEW

File: python/mxnet/gluon/block.py [hunk 1]
```diff
@@ -191,6 +191,8 @@ def __setattr__(self, name, value):
                 for i, c in enumerate(self._children):
                     if c is existing:
                         self._children[i] = value
+            if isinstance(value, Block):
```
Comment (piiswrong) [inline python/mxnet/gluon/block.py]: This should be elif. Otherwise value will be registered twice.

**Comment by piiswrong** [comment]
@szha

File: python/mxnet/gluon/block.py (hunk 1 above)
Comment (piiswrong) [reply python/mxnet/gluon/block.py]: Also there should be a break after 193

Comment (szha) [reply python/mxnet/gluon/block.py]: There's currently no guarantee that multiple children can't be the same instance, so not breaking is intentional.

Comment (szha) [reply python/mxnet/gluon/block.py]: Cleaner way would be to use OrderedDict for children.

**Comment by szha** [comment]
@leezu could you also include a fix from #8152 here? It is to add a `self._clear_cached_op()` call in `HybridBlock`'s `__setattr__`: https://github.com/apache/incubator-mxnet/pull/8152/files#diff-29da832c2145752f3906a2f71c7b63baR335

**Comment by leezu** [comment]
@szha the fix is included now.

**Comment by szha** [comment]
@leezu could you also include [this part of the test](https://github.com/apache/incubator-mxnet/pull/8152/files#diff-962bd5bb7248659d7eb3be37ee8a4c6bR519)? Thanks :)

**Comment by leezu** [comment]
@szha the tests are included now. Sorry, I overlooked that part of your message at first ;)

**Pull request merged**